### PR TITLE
Add operation tests on writeTexture() and B2T copies with stencil aspects

### PR DESCRIPTION
**Pull request by Jiawei-Shao**
<hr>

**Author checklist for test code/plans:**

- [ ] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [ ] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [ ] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: unit: checkStencilTextureContent

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:651]: nit: the separation between the function declaration and its body is a bit hard to see. Ditto below for the fragment shader.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: FYI @kainino0x maybe we want to move this texture helper to be shared between tests in the future.

Might be worth doing a perfect copy of the stencil data instead of checking each pixel individually below. It should be possible to do this with 8 draws with blendMode ADD that each check a single bit of the stencil buffer.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:718]: nit: mipLevelCount: 1 too I'd imagine.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: As a crazy hack we could use the instanceIndex to pass extra data with baseInstance without adding a uniform buffer ^^

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1281]: Sorry this is a big change, but wouldn't there be a way to add stencil8 to the kWorkingFormats and make the texture check use the new method you made, such that all the tests will automatically start testing stencil8?

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (kainino0x) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: > Might be worth doing a perfect copy of the stencil data instead of checking each pixel individually below. It should be possible to do this with 8 draws with blendMode ADD that each check a single bit of the stencil buffer.

This is one of the things I want to have a texture helper for.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1281]: Sorry I may not fully understand your comment here. In fact ```stencil8``` is included in ```kDepthStencilFormats``` and is tested in this test.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: As in WebGPU we support copying from the stencil aspect into a buffer, I think it easier to verify the stencil data with T2B copies. In this test I am testing the B2T and T2B copies, so I have to check the stencil data in a render pass, so it seems we just need to keep this code here.

Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: > Might be worth doing a perfect copy of the stencil data instead of checking each pixel individually below. It should be possible to do this with 8 draws with blendMode ADD that each check a single bit of the stencil buffer.

Do you mean using 8 render pipelines with different stencilReadMask (1 << 0 to 1<< 7), and in each draw we output ( (1<<0) / 255, (1 << 1) / 255, ...  (1 << 7) / 255) to an R8Unorm color output?

Comment (Kangz) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:630]: Yes exactly.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts]: Does it look better now?

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:652]: nit: I'm a bit surprised we're doing a texture to compute the bit -> r8unorm float value, when we could just output `(1 << param.stencilBitIndex) / 255` in the shader. Am I missing something?

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:684]: nit: I think there's a createBufferFromContents helper we can use.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:749]: nit: I think we might be able to skip alpha since we're r8unorm that has no alpha channel.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [inline src/webgpu/api/operation/command_buffer/image_copy.spec.ts:826]: u-nit: I think it might be more idiomatic JS to first create the full descriptor and continuously update it here. Especially since the rest of the GPUDepthStencilState doesn't change:

```
renderPipelineDescriptor.depthStencil.stencilReadMask = 1 << stencilBitIndex;
const renderPipeline = this.device.createRenderPipeline(renderPipelineDescriptor);
```

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1385]: Ah sorry I wasn't clear. In the file we have a lot of tests exercising many options of copies: bytesPerRow, rowsPerImage, sub copies, copies at an offset, etc. It would be really nice to find a way to "just" add the stencil copy tests to the parametrization of existing tests so that we get all the amazing coverage.

**Review by Kangz — COMMENTED** [review]
Stencil checking algorithm looks much better, thank you!

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:749]: Oh we cannot skip setting alpha as GPUBlendState.alpha is not optional.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1385]: Oh the reason I prefer adding a separated test is that the copies with depth/stencil formats are a bit different as we can only do full subresource copies with depth/stencil formats and the way to verify the result is totally different from other copies. What do you think?

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (kainino0x) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:713]: That's right. You can set it to `{}` though (if you want).

**Comment by kainino0x** [comment]
Please let me know if a second review is needed, I am skipping this for now to burn down my review inbox.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Kangz) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1356]: @kainino0x FYI this thread raises interesting points about helpers not handling the notion of aspects as consistently as they should.

Long term I think it would be better to have the tests be generic on the aspects etc, but that can be left for later. The solution you outlined with similar looking tests for various aspects being grouped together sounds good. (although it may be better to use `.expand(commonOffsetsAndSizeParamExpander)`). To make code duplication of code minimal, do you think we could have a helper like `uploadTextureAndVerifyCopy` but for depth / stencil?

All of the above said, I think we can land this to make forward progress first, then add the rest of the tests. This is a good step.

**Review by Kangz — APPROVED** [review]

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (kainino0x) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:634]:
```suggestion
  // TODO(crbug.com/dawn/868): Revisit this when consolidating texture helpers.
  async checkStencilTextureContent(
```

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (kainino0x) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1356]: The aspects problem is an issue with both the helpers and the capabilities tables. Agree we definitely need to fix it when fixing the helpers.

Also agree with using simple local helpers (just blocks of code extracted out into functions) to reduce code duplication, but it's not critical. Jiawei and I discussed this a bit over chat in the past few days as well.

> code duplication of code

"code duplication," i.e., duplication of the word "code" 👍 🤣

Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts:1356]: Thanks for your understanding, @kainino0x and @Kangz!

I plan to consider extracting the common code paths and better organizing all the buffer-to-texture copy tests after adding b2t-stencil-copy, t2b-stencil-copy and t2b-depth-copy.

**Comment by kainino0x** [comment]
@Jiawei-Shao this is ready to land right? The refactor will be a followup? (https://github.com/gpuweb/cts/pull/606#discussion_r670932971)

**Comment by Jiawei-Shao** [comment]
Hi @kainino0x, sorry that I am still preparing some changes locally and I will update them soon.  This PR is quite important to me as I need this PR to set up the basic infrastructure for other depth-stencil copy tests.

File: src/webgpu/api/operation/command_buffer/image_copy.spec.ts
Comment (Jiawei-Shao) [reply src/webgpu/api/operation/command_buffer/image_copy.spec.ts]: Hi @Kangz and @kainino0x, I've updated the PR. PTAL, thanks!

**Comment by Kangz** [comment]
Sorry I didn't get to re-review today. (Github doesn't send an email on just new commits I think). @kainino0x would you be able to take a look? Otherwise I'll do it tomorrow.

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
LGTM, please merge if ready

**Pull request merged**
